container_workflow_tool/distgit.py

```diff
@@ -371,39 +372,43 @@ def _handle_dangling_symlinks(self, src_parent, dest_parent):
-                    dest_target = os.path.join(os.path.dirname(dest_file), os.readlink(dest_file))
+                if dest_file.is_symlink():
+                    if os.path.isabs(os.readlink(str(dest_file))):
+                        self.logger.debug(f"Symlink {dest_file} contains absolute path. Skip it.")
+                        continue
+                    dest_target = dest_file.parents[0] / dest_file.readlink()
                     msg = f"looking for dangling symlink {dest_file} (that points to {dest_target})"
                     self.logger.debug(msg)
-                    if os.path.exists(dest_target):
+                    if dest_target.exists():
                         continue
                     # We found a dangling symlink to relative path,
                     # so we need to use the matching path in source,
                     # which means removing destination name
                     # from destination and adding it to source root
-                    dest_path_rel = re.sub(
-                        r"^{comp}{sep}".format(comp=dest_parent, sep=os.path.sep),
-                        "",
-                        dest_file
-                    )
-                    src_path_content = os.path.join(src_parent, dest_path_rel)
-                    self.logger.debug("unlink {dest}".format(dest=dest_file))
-                    os.unlink(dest_file)
-                    src_full = os.path.join(os.path.dirname(src_path_content),
-                                            os.readlink(src_path_content))
-                    if os.path.isdir(src_full):
+                    dest_path_rel = dest_file.relative_to(dest_parent)
+                    src_path_content = Path(src_parent) / dest_path_rel
+                    self.logger.debug(f"unlink {dest_file}")
+                    dest_file.unlink()
+                    src_full = src_path_content.parents[0] / src_path_content.readlink()
+                    if src_full.is_dir():
                         # In this case, when the source directory includes another symlinks outside
                         # of this directory, those wouldn't be fixed, so let's run the same function
                         # to fix dangling symlinks recursively.
                         self.logger.debug("cp -r {src} {dest}".format(src=src_full, dest=dest_file))
                         shutil.copytree(src_full, dest_file, symlinks=True)
                         self._handle_dangling_symlinks(src_parent, dest_parent)
                     else:
-                        self.logger.debug("cp {src} {dest}".format(src=src_full, dest=dest_file))
+                        if os.path.islink(src_full):
+                            self.logger.debug(f"Source file {src_full} is symlink.")
+                            self.logger.debug(f"Link is: {src_full.readlink()}")
+                            src_full = src_parent / src_full.readlink()
```

Comment:
Will this work? If the symlink is relative (the check for that is missing as well btw), then it will be relative to `src_full.parents[0]`, not `src_parent`.

Best thing would be if we could just make sure the function is called recursively on the directory the symlink lives in

Reply:
Note: `pathlib.Path.readlink` only exists in 3.9 and later, not sure if this is important.

Reply:
>Best thing would be if we could just make sure the function is called recursively on the directory the symlink lives in

Thinking a bit more on that topic, that would not help either, since the symlink would still be relative to the directory it lives in :/

Reply:
> Will this work? If the symlink is relative (the check for that is missing as well btw), then it will be relative to `src_full.parents[0]`, not `src_parent`.

I will check it. Good point, but I guess it works. But let me check it.